test/hotspot/jtreg/runtime/Monitor/libIterateMonitorWithDeadObjectTest.c

```diff
@@ -0,0 +1,142 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include <jni.h>
+#include <stdlib.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#define die(x) do { printf("%s:%s\n",x , __func__); perror(x); exit(EXIT_FAILURE); } while (0)
+
+#ifndef _Included_IterateMonitorWithDeadObjectTest
+#define _Included_IterateMonitorWithDeadObjectTest
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static JavaVM* jvm;
+static pthread_t attacher;
+
+static jobject create_object(JNIEnv* env) {
+  jclass clazz = (*env)->FindClass(env, "java/lang/Object");
+  if (clazz == 0) die("No class");
+
+  jmethodID constructor = (*env)->GetMethodID(env, clazz, "<init>", "()V");
+  if (constructor == 0) die("No constructor");
+
+  return (*env)->NewObject(env, clazz, constructor);
+}
+
+static void system_gc(JNIEnv* env) {
+  jclass clazz = (*env)->FindClass(env, "java/lang/System");
+  if (clazz == 0) die("No class");
+
+  jmethodID method = (*env)->GetStaticMethodID(env, clazz, "gc", "()V");
+  if (method == 0) die("No method");
+
+  (*env)->CallStaticVoidMethod(env, clazz, method);
+}
+
+static void thread_dump_with_locked_monitors(JNIEnv* env) {
+  jclass ManagementFactoryClass = (*env)->FindClass(env, "java/lang/management/ManagementFactory");
+  if (ManagementFactoryClass == 0) die("No ManagementFactory class");
+
+  jmethodID getThreadMXBeanMethod = (*env)->GetStaticMethodID(env, ManagementFactoryClass, "getThreadMXBean", "()Ljava/lang/management/ThreadMXBean;");
+  if (getThreadMXBeanMethod == 0) die("No getThreadMXBean method");
+
+  jobject threadBean = (*env)->CallStaticObjectMethod(env, ManagementFactoryClass, getThreadMXBeanMethod);
+
+  jclass ThreadMXBeanClass = (*env)->FindClass(env, "java/lang/management/ThreadMXBean");
+  if (ThreadMXBeanClass == 0) die("No ThreadMXBean class");
+
+  jmethodID dumpAllThreadsMethod = (*env)->GetMethodID(env, ThreadMXBeanClass, "dumpAllThreads", "(ZZ)[Ljava/lang/management/ThreadInfo;");
+  if (dumpAllThreadsMethod == 0) die("No dumpAllThreads method");
+
+  // The 'lockedMonitors == true' is what triggers the collection of the monitor with the dead object.
+  (*env)->CallObjectMethod(env, ThreadMXBeanClass, dumpAllThreadsMethod, threadBean, JNI_TRUE /* lockedMonitors */, JNI_FALSE /* lockedSynchronizers*/);
+}
+
+static void* do_test() {
+  JNIEnv* env;
+  int res = (*jvm)->AttachCurrentThread(jvm, (void**)&env, NULL);
+  if (res != JNI_OK) die("AttachCurrentThread");
+
+  jobject obj = create_object(env);
+
+  if ((*env)->MonitorEnter(env, obj) != 0) die("MonitorEnter");
+
+  // Drop the last strong reference to the object associated with the monitor.
+  // The monitor only keeps a weak reference to the object.
+  (*env)->DeleteLocalRef(env, obj);
+
+  // Let the GC clear the weak reference to the object.
+  system_gc(env);
```

Comment:
ZGC will clear the weak reference. G1 clears the weak reference, except for in a few specific situations. I've verified that G1 clears this weak reference, so I don't think it would be worth making this test longer or more complicated.